### PR TITLE
Changed Speed based Color Mapping from Theme Color to TurboColorRamp

### PR DESCRIPTION
**Pull request by Kalabint**
Thanks to the work of @jcardus the Replay and Report Lines are now colored based on the current speed.

What i noticed is, that the Information density with the MUI Color Scheme is not so great, because the second half is mapping from Orange to Red, which are very similar colors.

image

This is the reason, i'm proposing the [Turbo Color Ramp](https://research.google/blog/turbo-an-improved-rainbow-colormap-for-visualization/)  as an alternative.

image

Using the Turbo Color Ramp it would look something like this in Traccar:

image

_A Discussion point is, if the Scale should be dynamic, or based on a customizable Attribute (0-200km/h for example).
Benefit of the Attribute would be, that one can see at a glance, how fast a Vehicle is moving, without checking a Scale first._

### CODE REVIEW

**Comment by tananaev** [comment]
What I don't like is the hardcoded color palette. Is there a way to generate it?

**Comment by Kalabint** [comment]
Does the license text need to be in the file or is it ok to leave it out?

**Comment by tananaev** [comment]
You can probably leave it out. The new code looks much better, but make sure you run lint on it.

File: src/common/util/colors.js [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
+// Turbo Colormap
+const turboPolynomials = {
+  r: [0.13572138, 4.61539260, -42.66032258, 132.13108234, -152.94239396, 59.28637943],
+  g: [0.09140261, 2.19418839, 4.84296658, -14.18503333, 4.27729857, 2.82956604],
+  b: [0.10667330, 12.64194608, -60.58204836, 110.36276771, -89.90310912, 27.34824973],
+};
 
-  const c1 = decomposeColor(color1).values;
-  const c2 = decomposeColor(color2).values;
+function interpolateChannel(normalizedValue, coeffs) {
```
Comment (tananaev) [inline src/common/util/colors.js:8]: Can you please change functions to lambdas for consistency with other code.

File: src/common/util/colors.js [hunk 2]
```diff
@@ -1,23 +1,31 @@
...
+  ];
+}
 
+const getSpeedColor = (speed, maxSpeed) => {
+  const normalizedSpeed = Math.max(0, Math.min(1, speed / maxSpeed));
+  const [r, g, b] = interpolateTurbo(normalizedSpeed);
   return `rgb(${r}, ${g}, ${b})`;
 };
 
-export const getSpeedColor = (color1, color2, color3, speed, max) => {
```
[the remark this answers is not in the document]
Comment (Kalabint) [reply src/common/util/colors.js:17]: Reason for change to default export was the LINT error message

```
  25:1  error  Prefer default export on a file with single export
```

[the remark this answers is not in the document]
Comment (tananaev) [reply src/common/util/colors.js:17]: I guess we can keep it for now then.

**Comment by Kalabint** [comment]
Based on [this ](https://www.traccar.org/forums/topic/color-speed-replay/ )and my own needs, i added a custom attribute to set the max expected device speed.

Is this addition ok?

**Comment by tananaev** [comment]
Sorry, but it's not ok to add to this PR. Let's keep this as it was and send a separate PR if you want something else as well.

**Comment by Kalabint** [comment]
> Sorry, but it's not ok to add to this PR. Let's keep this as it was and send a separate PR if you want something else as well.

Taking things step by step is the right thing to do.

[Currently working on a Scale](https://github.com/traccar/traccar-web/pull/1243#issuecomment-2229314188), will create a new PR, when it becomes relevant, if at all, still have some work to do there.

image

**Comment by tananaev** [comment]
Merged, thank you.

**Pull request merged**